File: libs/server/infrastructure/src/pi_rpc_domain_architect.rs
use std::{path::PathBuf, process::Stdio, time::Duration};

use async_stream::try_stream;
use evidence_server_domain::{
    DomainArchitect, DomainArchitectEventStream, ModelingEvent, ServerError,
};
use serde_json::{json, Value};
use tokio::{
    io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader},
    process::Command,
    time,
};

const DEFAULT_TIMEOUT: Duration = Duration::from_secs(120);
const MODELING_TOOLS: &str = "read,edit,write,ls,find,grep";

#[derive(Debug, Clone)]
pub struct PiRpcDomainArchitectConfig {
    pub command: String,
    pub args: Vec<String>,
    pub timeout: Duration,
}

impl Default for PiRpcDomainArchitectConfig {
    fn default() -> Self {
        Self {
            command: "pi".to_string(),
            args: vec![
                "--mode".to_string(),
                "rpc".to_string(),
                "--no-session".to_string(),
                "--no-extensions".to_string(),
                "--no-skills".to_string(),
                "--no-prompt-templates".to_string(),
                "--no-themes".to_string(),
                "--no-context-files".to_string(),
                "--tools".to_string(),
                MODELING_TOOLS.to_string(),
            ],
            timeout: DEFAULT_TIMEOUT,
        }
    }
}

fn workspace_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../..")
        .canonicalize()
        .unwrap_or_else(|_| PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../.."))
}

fn evidence_modeling_dir() -> PathBuf {
    let path = workspace_root().join(".evidence");
    path.canonicalize().unwrap_or(path)
}

#[derive(Debug, Clone, Default)]
pub struct PiRpcDomainArchitect {
    config: PiRpcDomainArchitectConfig,
}

impl PiRpcDomainArchitect {
    pub fn new(config: PiRpcDomainArchitectConfig) -> Self {
        Self { config }
    }

    pub fn with_command(command: impl Into<String>) -> Self {
        Self::new(PiRpcDomainArchitectConfig {
            command: command.into(),
            ..PiRpcDomainArchitectConfig::default()
        })
    }
}

impl DomainArchitect for PiRpcDomainArchitect {
    fn propose_model_stream(&self, requirement: String) -> DomainArchitectEventStream {
        let config = self.config.clone();

        Box::pin(try_stream! {
            let message = requirement;
            let mut command = Command::new(&config.command);
            command
                .args(&config.args)
                .current_dir(evidence_modeling_dir())
                .stdin(Stdio::piped())
                .stdout(Stdio::piped())
                .stderr(Stdio::piped())
                .kill_on_drop(true);

            let mut child = command.spawn().map_err(|error| {
                ServerError::Internal(format!("failed to start pi rpc process: {error}"))
            })?;

            let mut stdin = child
                .stdin
                .take()
                .ok_or_else(|| ServerError::Internal("pi rpc stdin unavailable".to_string()))?;
            let stdout = child
                .stdout
                .take()
                .ok_or_else(|| ServerError::Internal("pi rpc stdout unavailable".to_string()))?;
            let mut stderr = child
                .stderr
                .take()
                .ok_or_else(|| ServerError::Internal("pi rpc stderr unavailable".to_string()))?;

            let rpc_command = json!({
                "id": "evidence-chat",
                "type": "prompt",
                "message": message,
            });
            stdin
                .write_all(rpc_command.to_string().as_bytes())
                .await
                .map_err(|error| {
                    ServerError::Internal(format!("failed to write pi rpc message: {error}"))
                })?;
            stdin.write_all(b"\n").await.map_err(|error| {
                ServerError::Internal(format!("failed to write pi rpc message: {error}"))
            })?;
            stdin.flush().await.map_err(|error| {
                ServerError::Internal(format!("failed to flush pi rpc message: {error}"))
            })?;

            let mut lines = BufReader::new(stdout).lines();
            let mut assistant_text = String::new();
            let mut accepted = false;
            let mut saw_message_end = false;
            let mut saw_agent_end = false;
            let mut saw_agent_settled = false;

            loop {
                let maybe_line = time::timeout(config.timeout, lines.next_line())
                    .await
                    .map_err(|_| ServerError::Internal("pi rpc request timed out".to_string()))?
                    .map_err(|error| {
                        ServerError::Internal(format!("failed to read pi rpc output: {error}"))
                    })?;
                let Some(line) = maybe_line else {
                    break;
                };

                if line.trim().is_empty() {
                    continue;
                }

                let event: Value = serde_json::from_str(&line).map_err(|error| {
                    ServerError::Internal(format!("failed to parse pi rpc JSON line: {error}"))
                })?;

                match event.get("type").and_then(Value::as_str) {
                    Some("response") => {
                        if event.get("command").and_then(Value::as_str) == Some("prompt") {
                            if event.get("success").and_then(Value::as_bool) == Some(false) {
                                let error = event
                                    .get("error")
                                    .and_then(Value::as_str)
                                    .unwrap_or("pi rpc message rejected");
                                let _ = child.kill().await;
                                Err(ServerError::Internal(error.to_string()))?;
                            }
                            accepted = true;
                        }
                    }
                    Some("message_update") => {
                        let assistant_event = event
                            .get("assistantMessageEvent")
                            .unwrap_or(&Value::Null);
                        match assistant_event.get("type").and_then(Value::as_str) {
                            Some("text_delta") => {
                                if let Some(delta) = assistant_event.get("delta").and_then(Value::as_str) {
                                    assistant_text.push_str(delta);
                                    yield ModelingEvent::TextChunk { chunk: delta.to_string() };
                                }
                            }
                            Some("thinking_start") => {
                                yield ModelingEvent::ReasoningStarted;
                            }
                            Some("thinking_delta") => {
                                if let Some(delta) = assistant_event.get("delta").and_then(Value::as_str) {
                                    yield ModelingEvent::ReasoningChunk { chunk: delta.to_string() };
                                }
                            }
                            Some("thinking_end") => {
                                yield ModelingEvent::ReasoningEnded;
                            }
                            Some("toolcall_start") => {
                                let tool_call_id = tool_call_id(assistant_event)?;
                                let tool_name = tool_name(assistant_event);
                                yield ModelingEvent::ToolCallStarted { tool_call_id, tool_name };
                            }
                            Some("toolcall_delta") => {
                                if let Some(delta) = assistant_event.get("delta").and_then(Value::as_str) {
                                    let tool_call_id = tool_call_id(assistant_event)?;
                                    let tool_name = tool_name(assistant_event);
                                    yield ModelingEvent::ToolCallDelta {
                                        tool_call_id,
                                        tool_name,
                                        chunk: delta.to_string(),
                                    };
                                }
                            }
                            Some("toolcall_end") => {
                                let tool_call = assistant_tool_call(assistant_event)?;
                                let tool_call_id = required_string(tool_call, "id", "pi rpc tool call")?;
                                let tool_name = required_string(tool_call, "name", "pi rpc tool call")?;
                                let input = required_value(tool_call, "arguments", "pi rpc tool call")?;
                                yield ModelingEvent::ToolCallReady { tool_call_id, tool_name, input };
                            }
                            _ => {}
                        }
                    }
                    Some("tool_execution_start") => {
                        yield ModelingEvent::ToolExecutionStarted {
                            tool_call_id: required_event_string(&event, "toolCallId")?,
                            tool_name: required_event_string(&event, "toolName")?,
                            args: required_event_value(&event, "args")?,
                        };
                    }
                    Some("tool_execution_update") => {
                        yield ModelingEvent::ToolExecutionUpdated {
                            tool_call_id: required_event_string(&event, "toolCallId")?,
                            tool_name: required_event_string(&event, "toolName")?,
                            args: required_event_value(&event, "args")?,
                            partial_result: required_event_value(&event, "partialResult")?,
                        };
                    }
                    Some("tool_execution_end") => {
                        let tool_call_id = required_event_string(&event, "toolCallId")?;
                        let tool_name = required_event_string(&event, "toolName")?;
                        let result = required_event_value(&event, "result")?;
                        let is_error = event
                            .get("isError")
                            .and_then(Value::as_bool)
                            .ok_or_else(|| ServerError::Internal("pi rpc isError missing".to_string()))?;
                        yield ModelingEvent::ToolExecutionEnded {
                            tool_call_id,
                            tool_name,
                            result,
                            is_error,
                        };
                    }
                    Some("message_end") => {
                        let message = event.get("message").unwrap_or(&Value::Null);
                        if message.get("role").and_then(Value::as_str) == Some("assistant") {
                            saw_message_end = true;
                            if assistant_text.trim().is_empty() {
                                let text = extract_message_text(message);
                                if !text.is_empty() {
                                    assistant_text.push_str(&text);
                                    yield ModelingEvent::TextChunk { chunk: text };
                                }
                            }
                        }
                    }
                    Some("agent_end") => {
                        saw_agent_end = true;
                        if assistant_text.trim().is_empty() {
                            let text = extract_agent_end_text(&event);
                            if !text.is_empty() {
                                assistant_text.push_str(&text);
                                yield ModelingEvent::TextChunk { chunk: text };
                            }
                        }
                    }
                    Some("agent_settled") => {
                        saw_agent_settled = true;
                        break;
                    }
                    _ => {}
                }
            }

            let _ = child.kill().await;
            let mut stderr_output = String::new();
            let _ = time::timeout(
                Duration::from_secs(1),
                stderr.read_to_string(&mut stderr_output),
            )
            .await;

            if !accepted {
                Err(ServerError::Internal(format!(
                    "pi rpc process ended before accepting message{}",
                    stderr_detail(&stderr_output)
                )))?;
            }
            if !saw_agent_settled {
                Err(ServerError::Internal(format!(
                    "pi rpc process ended before the agent settled{}",
                    stderr_detail(&stderr_output)
                )))?;
            }

            if saw_message_end {
                yield ModelingEvent::MessageEnded;
            }
            if saw_agent_end {
                yield ModelingEvent::AgentEnded;
            }
            yield ModelingEvent::Completed;
        })
    }
}

fn assistant_tool_call(assistant_event: &Value) -> Result<&Value, ServerError> {
    if let Some(tool_call) = assistant_event.get("toolCall") {
        return Ok(tool_call);
    }

    assistant_event
        .get("contentIndex")
        .and_then(Value::as_u64)
        .and_then(|content_index| {
            assistant_event
                .get("partial")
                .and_then(|partial| partial.get("content"))
                .and_then(Value::as_array)
                .and_then(|content| content.get(content_index as usize))
        })
        .ok_or_else(|| ServerError::Internal("pi rpc tool call missing".to_string()))
}

fn tool_call_id(assistant_event: &Value) -> Result<String, ServerError> {
    required_string(
        assistant_tool_call(assistant_event)?,
        "id",
        "pi rpc tool call",
    )
}

fn tool_name(assistant_event: &Value) -> Option<String> {
    assistant_tool_call(assistant_event)
        .ok()
        .and_then(|tool_call| tool_call.get("name"))
        .and_then(Value::as_str)
        .map(str::to_string)
}

fn required_string(value: &Value, key: &str, subject: &str) -> Result<String, ServerError> {
    value
        .get(key)
        .and_then(Value::as_str)
        .map(str::to_string)
        .ok_or_else(|| ServerError::Internal(format!("{subject} {key} missing")))
}

fn required_value(value: &Value, key: &str, subject: &str) -> Result<Value, ServerError> {
    value
        .get(key)
        .cloned()
        .ok_or_else(|| ServerError::Internal(format!("{subject} {key} missing")))
}

fn required_event_string(event: &Value, key: &str) -> Result<String, ServerError> {
    required_string(event, key, "pi rpc event")
}

fn required_event_value(event: &Value, key: &str) -> Result<Value, ServerError> {
    required_value(event, key, "pi rpc event")
}

fn stderr_detail(stderr: &str) -> String {
    let trimmed = stderr.trim();
    if trimmed.is_empty() {
        String::new()
    } else {
        format!(": {trimmed}")
    }
}

fn extract_agent_end_text(event: &Value) -> String {
    event
        .get("messages")
        .and_then(Value::as_array)
        .and_then(|messages| {
            messages
                .iter()
                .rev()
                .find(|message| message.get("role").and_then(Value::as_str) == Some("assistant"))
        })
        .map(extract_message_text)
        .unwrap_or_default()
}

fn extract_message_text(message: &Value) -> String {
    match message.get("content") {
        Some(Value::String(text)) => text.clone(),
        Some(Value::Array(parts)) => parts
            .iter()
            .filter_map(|part| {
                if part.get("type").and_then(Value::as_str) == Some("text") {
                    part.get("text").and_then(Value::as_str)
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
            .join(""),
        _ => String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_config_enables_modeling_tools_without_preset_prompt() {
        let config = PiRpcDomainArchitectConfig::default();

        assert!(!config.args.iter().any(|arg| arg == "--no-tools"));
        assert!(!config
            .args
            .iter()
            .any(|arg| arg == "--append-system-prompt"));
        let tools_index = config
            .args
            .iter()
            .position(|arg| arg == "--tools")
            .expect("default config should constrain enabled tools");
        assert_eq!(
            config.args.get(tools_index + 1).map(String::as_str),
            Some(MODELING_TOOLS)
        );
    }

    #[test]
    fn reads_current_tool_call_event_shapes() {
        let partial = json!({
            "type": "toolcall_delta",
            "contentIndex": 0,
            "partial": {
                "content": [{
                    "type": "toolCall",
                    "id": "call-1",
                    "name": "edit",
                    "arguments": {"path": ".evidence/entities/order.yaml"}
                }]
            }
        });
        let complete = json!({
            "type": "toolcall_end",
            "toolCall": {
                "type": "toolCall",
                "id": "call-1",
                "name": "edit",
                "arguments": {"path": ".evidence/entities/order.yaml"}
            }
        });

        assert_eq!(tool_call_id(&partial).unwrap(), "call-1");
        assert_eq!(tool_name(&partial).as_deref(), Some("edit"));
        assert_eq!(tool_call_id(&complete).unwrap(), "call-1");
        assert_eq!(tool_name(&complete).as_deref(), Some("edit"));
    }

    #[test]
    fn rejects_retired_top_level_tool_call_fields() {
        let retired = json!({
            "type": "toolcall_start",
            "toolCallId": "call-1",
            "toolName": "edit"
        });

        assert!(tool_call_id(&retired).is_err());
        assert_eq!(tool_name(&retired), None);
    }

    #[test]
    fn extracts_text_from_agent_end_event() {
        let event = json!({
            "type": "agent_end",
            "messages": [
                {"role": "user", "content": "hello"},
                {"role": "assistant", "content": [{"type": "text", "text": "{\"summary\":null}"}]}
            ]
        });

        assert_eq!(extract_agent_end_text(&event), "{\"summary\":null}");
    }
}

File: libs/server-nest/infrastructure/src/lib/pi-rpc-domain-architect.ts
import { spawn, type ChildProcessWithoutNullStreams } from 'node:child_process';
import { StringDecoder } from 'node:string_decoder';
import {
  DomainError,
  type DomainArchitect,
  type ModelingEvent,
  type ModelingRequest,
} from '@evidence/server-nest-domain';

const DEFAULT_TIMEOUT_MS = 120_000;
const DEFAULT_ARGS = [
  '--mode',
  'rpc',
  '--no-session',
  '--no-approve',
  '--no-extensions',
  '--no-skills',
  '--no-prompt-templates',
  '--no-themes',
  '--no-context-files',
  '--tools',
  'read,edit,write,ls,find,grep',
] as const;
const MAX_STDERR_LENGTH = 8_192;

export interface PiRpcDomainArchitectOptions {
  command?: string;
  args?: readonly string[];
  timeoutMs?: number;
  env?: NodeJS.ProcessEnv;
}

interface RpcState {
  accepted: boolean;
  assistantMessageIndex: number | null;
  assistantText: string;
  sawMessageEnd: boolean;
  sawAgentEnd: boolean;
}

interface ProcessExit {
  code: number | null;
  signal: NodeJS.Signals | null;
}

export class PiRpcDomainArchitect implements DomainArchitect {
  private readonly command: string;
  private readonly args: readonly string[];
  private readonly timeoutMs: number;
  private readonly env: NodeJS.ProcessEnv;

  constructor(options: PiRpcDomainArchitectOptions = {}) {
    this.command =
      options.command ?? process.env['EVIDENCE_PI_COMMAND'] ?? 'pi';
    this.args = options.args ?? DEFAULT_ARGS;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.env = {
      ...process.env,
      PI_SKIP_VERSION_CHECK: '1',
      ...options.env,
    };
  }

  async *proposeModelStream(
    request: ModelingRequest,
  ): AsyncIterable<ModelingEvent> {
    const child = this.startProcess(request.modelDirectory);
    let stderr = '';
    child.stderr.setEncoding('utf8');
    child.stderr.on('data', (chunk: string) => {
      stderr = `${stderr}${chunk}`.slice(-MAX_STDERR_LENGTH);
    });

    const exit = processExit(child);
    const failure = processFailure(child, this.timeoutMs, request.signal);
    const state: RpcState = {
      accepted: false,
      assistantMessageIndex: null,
      assistantText: '',
      sawMessageEnd: false,
      sawAgentEnd: false,
    };

    try {
      await Promise.race([
        writeCommand(child, {
          id: 'evidence-chat',
          type: 'prompt',
          message: request.requirement,
        }),
        failure.promise,
      ]);

      for await (const message of jsonLines(child, failure.promise)) {
        for (const event of mapRpcMessage(message, state)) {
          yield event;
        }
        if (isSettled(message)) {
          return;
        }
      }

      const result = await Promise.race([exit, failure.promise]);
      if (result.code !== 0) {
        throw processExitError(result, stderr);
      }
      if (!state.accepted) {
        throw processEndedError(
          'Pi RPC ended before accepting the modeling request',
          stderr,
        );
      }
      throw processEndedError('Pi RPC ended before the agent settled', stderr);
    } catch (error) {
      if (error instanceof DomainError) {
        throw error;
      }
      throw DomainError.internal(errorMessage(error));
    } finally {
      failure.dispose();
      await stopProcess(child, exit);
    }
  }

  private startProcess(modelDirectory: string): ChildProcessWithoutNullStreams {
    return spawn(this.command, [...this.args], {
      cwd: modelDirectory,
      env: this.env,
      stdio: ['pipe', 'pipe', 'pipe'],
      windowsHide: true,
    });
  }
}

function processFailure(
  child: ChildProcessWithoutNullStreams,
  timeoutMs: number,
  signal?: AbortSignal,
): { promise: Promise<never>; dispose: () => void } {
  let rejectFailure: (error: Error) => void = () => undefined;
  const promise = new Promise<never>((_resolve, reject) => {
    rejectFailure = reject;
  });
  const onError = (error: Error) => {
    rejectFailure(
      DomainError.internal(`failed to start Pi RPC: ${error.message}`),
    );
  };
  const onAbort = () => {
    rejectFailure(DomainError.internal('Pi RPC request was aborted'));
    child.kill('SIGTERM');
  };
  const timer = setTimeout(() => {
    rejectFailure(
      DomainError.internal(`Pi RPC request timed out after ${timeoutMs}ms`),
    );
    child.kill('SIGTERM');
  }, timeoutMs);

  child.once('error', onError);
  signal?.addEventListener('abort', onAbort, { once: true });
  if (signal?.aborted) {
    onAbort();
  }

  return {
    promise,
    dispose: () => {
      clearTimeout(timer);
      child.off('error', onError);
      signal?.removeEventListener('abort', onAbort);
    },
  };
}

function processExit(
  child: ChildProcessWithoutNullStreams,
): Promise<ProcessExit> {
  return new Promise((resolve) => {
    child.once('close', (code, signal) => resolve({ code, signal }));
  });
}

async function writeCommand(
  child: ChildProcessWithoutNullStreams,
  command: unknown,
): Promise<void> {
  await new Promise<void>((resolve, reject) => {
    child.stdin.write(
      `${JSON.stringify(command)}\n`,
      (error?: Error | null) => {
        if (error) {
          reject(
            DomainError.internal(
              `failed to write Pi RPC command: ${error.message}`,
            ),
          );
        } else {
          resolve();
        }
      },
    );
  });
}

async function* jsonLines(
  child: ChildProcessWithoutNullStreams,
  failure: Promise<never>,
): AsyncIterable<unknown> {
  const decoder = new StringDecoder('utf8');
  const iterator = child.stdout[Symbol.asyncIterator]();
  let buffer = '';

  while (true) {
    const next = await Promise.race([iterator.next(), failure]);
    if (next.done) {
      break;
    }
    buffer += decoder.write(
      Buffer.isBuffer(next.value) ? next.value : Buffer.from(next.value),
    );

    let newline = buffer.indexOf('\n');
    while (newline >= 0) {
      const line = buffer.slice(0, newline).trim();
      buffer = buffer.slice(newline + 1);
      if (line.length > 0) {
        yield parseRpcLine(line);
      }
      newline = buffer.indexOf('\n');
    }
  }

  buffer += decoder.end();
  const finalLine = buffer.trim();
  if (finalLine.length > 0) {
    yield parseRpcLine(finalLine);
  }
}

function parseRpcLine(line: string): unknown {
  try {
    return JSON.parse(line) as unknown;
  } catch (error) {
    throw DomainError.internal(
      `Pi RPC emitted invalid JSON: ${errorMessage(error)}`,
    );
  }
}

function mapRpcMessage(message: unknown, state: RpcState): ModelingEvent[] {
  const record = asRecord(message);
  if (!record) {
    return [];
  }

  switch (stringField(record, 'type')) {
    case 'message_start':
      return mapMessageStart(record, state);
    case 'message_update':
      return mapMessageUpdate(record, state);
    case 'message_end':
      return mapMessageEnd(record, state);
    case 'agent_end':
      return mapAgentEnd(record, state);
    case 'tool_execution_start':
      return [mapToolExecutionStarted(record)];
    case 'tool_execution_update':
      return [mapToolExecutionUpdated(record)];
    case 'tool_execution_end':
      return [mapToolExecutionEnded(record)];
    case 'response':
      return mapResponse(record, state);
    case 'error':
    case 'extension_error':
      throw DomainError.internal(rpcError(record));
    case 'agent_settled':
    case 'settled':
      return mapAgentSettled(state);
    default:
      return [];
  }
}

function mapMessageStart(
  record: Record<string, unknown>,
  state: RpcState,
): ModelingEvent[] {
  const message = asRecord(record['message']);
  if (stringField(message, 'role') !== 'assistant') {
    return [];
  }
  state.assistantMessageIndex = numberField(record, 'messageIndex');
  return [];
}

function mapMessageUpdate(
  record: Record<string, unknown>,
  state: RpcState,
): ModelingEvent[] {
  const messageIndex = numberField(record, 'messageIndex');
  if (
    state.assistantMessageIndex !== null &&
    messageIndex !== null &&
    state.assistantMessageIndex !== messageIndex
  ) {
    return [];
  }

  const event = asRecord(record['assistantMessageEvent']);
  if (!event) {
    return [];
  }
  const eventType = stringField(event, 'type');
  const chunk =
    stringField(event, 'delta') ?? stringField(event, 'content') ?? '';

  switch (eventType) {
    case 'text_delta':
      state.assistantText += chunk;
      return [{ type: 'text-chunk', chunk }];
    case 'thinking_start':
      return [{ type: 'reasoning-started' }];
    case 'thinking_delta':
      return [{ type: 'reasoning-chunk', chunk }];
    case 'thinking_end':
      return [{ type: 'reasoning-ended' }];
    case 'toolcall_start':
      return [mapToolCallStarted(event)];
    case 'toolcall_delta':
      return [mapToolCallDelta(event)];
    case 'toolcall_end':
      return [mapToolCallReady(event)];
    default:
      return [];
  }
}

function mapMessageEnd(
  record: Record<string, unknown>,
  state: RpcState,
): ModelingEvent[] {
  const message = asRecord(record['message']);
  if (stringField(message, 'role') !== 'assistant') {
    return [];
  }
  state.sawMessageEnd = true;
  return fallbackText(extractMessageText(message), state);
}

function mapAgentEnd(
  record: Record<string, unknown>,
  state: RpcState,
): ModelingEvent[] {
  state.sawAgentEnd = true;
  const messages = Array.isArray(record['messages']) ? record['messages'] : [];
  const assistant = [...messages]
    .reverse()
    .map(asRecord)
    .find((message) => stringField(message, 'role') === 'assistant');
  return fallbackText(extractMessageText(assistant ?? null), state);
}

function fallbackText(text: string, state: RpcState): ModelingEvent[] {
  if (state.assistantText.trim().length > 0 || text.length === 0) {
    return [];
  }
  state.assistantText += text;
  return [{ type: 'text-chunk', chunk: text }];
}

function extractMessageText(message: Record<string, unknown> | null): string {
  const content = message?.['content'];
  if (typeof content === 'string') {
    return content;
  }
  if (!Array.isArray(content)) {
    return '';
  }
  return content
    .map(asRecord)
    .filter((part) => stringField(part, 'type') === 'text')
    .map((part) => stringField(part, 'text') ?? '')
    .join('');
}

function mapToolCallStarted(event: Record<string, unknown>): ModelingEvent {
  const toolCall = assistantToolCall(event);
  return {
    type: 'tool-call-started',
    toolCallId: requiredString(toolCall, 'id', 'Pi RPC tool call'),
    toolName: stringField(toolCall, 'name'),
  };
}

function mapToolCallDelta(event: Record<string, unknown>): ModelingEvent {
  const toolCall = assistantToolCall(event);
  return {
    type: 'tool-call-delta',
    toolCallId: requiredString(toolCall, 'id', 'Pi RPC tool call'),
    toolName: stringField(toolCall, 'name'),
    chunk: requiredString(event, 'delta', 'Pi RPC tool call delta'),
  };
}

function mapToolCallReady(event: Record<string, unknown>): ModelingEvent {
  const toolCall = assistantToolCall(event);
  return {
    type: 'tool-call-ready',
    toolCallId: requiredString(toolCall, 'id', 'Pi RPC tool call'),
    toolName: requiredString(toolCall, 'name', 'Pi RPC tool call'),
    input: requiredValue(toolCall, 'arguments', 'Pi RPC tool call'),
  };
}

function assistantToolCall(
  event: Record<string, unknown>,
): Record<string, unknown> {
  const direct = asRecord(event['toolCall']) ?? asRecord(event['partialCall']);
  if (direct) {
    return direct;
  }

  const contentIndex = numberField(event, 'contentIndex');
  const partial = asRecord(event['partial']);
  const content = partial?.['content'];
  const indexed =
    contentIndex !== null && Array.isArray(content)
      ? asRecord(content[contentIndex])
      : null;
  if (!indexed) {
    throw DomainError.internal('Pi RPC tool call missing');
  }
  return indexed;
}

function mapToolExecutionStarted(
  record: Record<string, unknown>,
): ModelingEvent {
  return {
    type: 'tool-execution-started',
    toolCallId: requiredString(record, 'toolCallId', 'Pi RPC event'),
    toolName: requiredString(record, 'toolName', 'Pi RPC event'),
    args: requiredValue(record, 'args', 'Pi RPC event'),
  };
}

function mapToolExecutionUpdated(
  record: Record<string, unknown>,
): ModelingEvent {
  return {
    type: 'tool-execution-updated',
    toolCallId: requiredString(record, 'toolCallId', 'Pi RPC event'),
    toolName: requiredString(record, 'toolName', 'Pi RPC event'),
    args: requiredValue(record, 'args', 'Pi RPC event'),
    partialResult: requiredValue(record, 'partialResult', 'Pi RPC event'),
  };
}

function mapToolExecutionEnded(record: Record<string, unknown>): ModelingEvent {
  if (typeof record['isError'] !== 'boolean') {
    throw DomainError.internal('Pi RPC event isError missing');
  }
  return {
    type: 'tool-execution-ended',
    toolCallId: requiredString(record, 'toolCallId', 'Pi RPC event'),
    toolName: requiredString(record, 'toolName', 'Pi RPC event'),
    result: requiredValue(record, 'result', 'Pi RPC event'),
    isError: record['isError'],
  };
}

function mapResponse(
  record: Record<string, unknown>,
  state: RpcState,
): ModelingEvent[] {
  if (stringField(record, 'command') !== 'prompt') {
    return [];
  }
  if (record['success'] === false) {
    throw DomainError.internal(rpcError(record));
  }
  state.accepted = true;
  return [];
}

function mapAgentSettled(state: RpcState): ModelingEvent[] {
  if (!state.accepted) {
    throw DomainError.internal(
      'Pi RPC settled before accepting the modeling request',
    );
  }
  const events: ModelingEvent[] = [];
  if (state.sawMessageEnd) {
    events.push({ type: 'message-ended' });
  }
  if (state.sawAgentEnd) {
    events.push({ type: 'agent-ended' });
  }
  events.push({ type: 'completed' });
  return events;
}

function rpcError(record: Record<string, unknown>): string {
  const error = asRecord(record['error']);
  return (
    stringField(error, 'message') ??
    (typeof record['error'] === 'string' ? record['error'] : null) ??
    stringField(record, 'message') ??
    'Pi RPC request failed'
  );
}

function requiredString(
  record: Record<string, unknown>,
  name: string,
  subject: string,
): string {
  const value = stringField(record, name);
  if (value === null) {
    throw DomainError.internal(`${subject} ${name} missing`);
  }
  return value;
}

function requiredValue(
  record: Record<string, unknown>,
  name: string,
  subject: string,
): unknown {
  if (!(name in record)) {
    throw DomainError.internal(`${subject} ${name} missing`);
  }
  return record[name];
}

function isSettled(message: unknown): boolean {
  const type = stringField(asRecord(message), 'type');
  return type === 'agent_settled' || type === 'settled';
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function stringField(
  record: Record<string, unknown> | null,
  name: string,
): string | null {
  const value = record?.[name];
  return typeof value === 'string' ? value : null;
}

function numberField(
  record: Record<string, unknown> | null,
  name: string,
): number | null {
  const value = record?.[name];
  return typeof value === 'number' ? value : null;
}

function processEndedError(message: string, stderr: string): DomainError {
  const detail = stderr.trim();
  return DomainError.internal(
    detail.length > 0 ? `${message}: ${detail}` : message,
  );
}

function processExitError(result: ProcessExit, stderr: string): DomainError {
  const detail = stderr.trim();
  const status = result.signal
    ? `signal ${result.signal}`
    : `status ${String(result.code)}`;
  return DomainError.internal(
    detail.length > 0
      ? `Pi RPC exited with ${status}: ${detail}`
      : `Pi RPC exited with ${status}`,
  );
}

async function stopProcess(
  child: ChildProcessWithoutNullStreams,
  exit: Promise<ProcessExit>,
): Promise<void> {
  child.stdin.destroy();
  if (child.exitCode !== null || child.signalCode !== null) {
    await exit;
    return;
  }

  child.kill('SIGTERM');
  const stopped = await Promise.race([
    exit.then(() => true),
    new Promise<false>((resolve) => setTimeout(() => resolve(false), 250)),
  ]);
  if (!stopped && child.exitCode === null && child.signalCode === null) {
    child.kill('SIGKILL');
    await exit;
  }
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
